File: shared/contracts/python/audit_event.py
"""
AuditEvent contract — Python / Pydantic.
Represents an immutable audit record. Once written, never modified.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel

from .governance_decision import GovernanceOutcome


class AuditEventType(str, Enum):
    # Session
    USER_LOGIN = "user.login"
    USER_LOGOUT = "user.logout"
    SESSION_EXPIRED = "user.session_expired"

    # Claims
    CLAIM_SELECTED = "claim.selected"
    CLAIM_SEARCHED = "claim.searched"

    # AI — reads
    SUMMARY_GENERATED = "ai.summary.generated"
    CONVERSATION_TURN_COMPLETED = "ai.conversation.turn_completed"
    KNOWLEDGE_RETRIEVED = "ai.rag.retrieved"

    # AI — writes
    NOTE_DRAFTED = "ai.note.drafted"

    # Governance
    GOVERNANCE_EVALUATED = "governance.evaluated"

    # Human decisions
    HUMAN_APPROVAL_REQUESTED = "human.approval.requested"
    HUMAN_APPROVAL_GRANTED = "human.approval.granted"
    HUMAN_APPROVAL_REJECTED = "human.approval.rejected"

    # Writes to external systems
    CLAIM_NOTE_WRITTEN = "claimcenter.note.written"
    CLAIM_NOTE_WRITE_FAILED = "claimcenter.note.write_failed"


class AuditEvent(BaseModel):
    event_id: str                    # UUID, assigned by audit store
    event_type: AuditEventType
    occurred_at: datetime            # UTC

    # Who and where
    user_id: str
    user_role: str
    session_id: str

    # What claim (if applicable)
    claim_id: str | None = None
    claim_number: str | None = None

    # Event-specific payload
    payload: dict[str, Any] = {}

    # Governance context
    governance_decision_id: str | None = None
    governance_outcome: GovernanceOutcome | None = None

    # Model context
    model_id: str | None = None
    task_type: str | None = None
    input_tokens: int | None = None
    output_tokens: int | None = None
    latency_ms: int | None = None

File: shared/contracts/typescript/auditEvent.ts
// AuditEvent contract
// Represents an immutable audit record. Once written, never modified.

import type { GovernanceOutcome } from "./governanceDecision";

export enum AuditEventType {
  // Session
  UserLogin = "user.login",
  UserLogout = "user.logout",
  SessionExpired = "user.session_expired",

  // Claims
  ClaimSelected = "claim.selected",
  ClaimSearched = "claim.searched",

  // AI — reads
  SummaryGenerated = "ai.summary.generated",
  ConversationTurnCompleted = "ai.conversation.turn_completed",
  KnowledgeRetrieved = "ai.rag.retrieved",

  // AI — writes
  NoteDrafted = "ai.note.drafted",

  // Governance
  GovernanceEvaluated = "governance.evaluated",

  // Human decisions
  HumanApprovalRequested = "human.approval.requested",
  HumanApprovalGranted = "human.approval.granted",
  HumanApprovalRejected = "human.approval.rejected",

  // Writes to external systems
  ClaimNoteWritten = "claimcenter.note.written",
  ClaimNoteWriteFailed = "claimcenter.note.write_failed",
}

export interface AuditEvent {
  eventId: string;               // UUID, assigned by audit store
  eventType: AuditEventType;
  occurredAt: string;            // ISO 8601 UTC

  // Who and where
  userId: string;
  userRole: string;
  sessionId: string;

  // What claim (if applicable)
  claimId: string | null;
  claimNumber: string | null;

  // Event-specific payload (varies by eventType)
  payload: Record<string, unknown>;

  // Governance context (if a governance decision was involved)
  governanceDecisionId: string | null;
  governanceOutcome: GovernanceOutcome | null;

  // Model context (if an AI model was called)
  modelId: string | null;
  taskType: string | null;
  inputTokens: number | null;
  outputTokens: number | null;
  latencyMs: number | null;
}
